### PR TITLE
Add Preview for LensFlare {Image, Circle and Polygon}

### PR DESCRIPTION
**Pull request by skhiat**
Add Lens Flare preview:

image

image
image
image

---
### Testing status
UI show only the thumbnail for LensFlare Circle and Polygon when it's collapsed.

Stretch:
https://docs.google.com/document/d/1AFL10SqfHy_rMv-mR9mccV0P7nxWhxd2-vK_X5ry7yM/edit#heading=h.offjzv72ayms

Test:
Image thumbnail with large ratio 'Scale': 1x50 and 1x100, 1x1000 should have the same thumbnail

### CODE REVIEW

**Review by JulienIgnace-Unity — APPROVED** [review]

**Review by iM0ve — CHANGES_REQUESTED** [review]
The PR quite simple, but found one case where it does not work. 
In the video, you can see I collapse the yellow circle element. This somehow messes up the ordering and the yellow circle thumbnail is now displayed for the second element. This is incorrect.


https://user-images.githubusercontent.com/36502659/118633307-0041d500-b7da-11eb-83e9-d83bea376985.mp4

**Comment by SeanPuller** [comment]
Good work, it gives a much better indication of the elements

Bugs
1. Polygon Thumbnail Issues

https://user-images.githubusercontent.com/43151199/118962983-40ce5980-b966-11eb-9c4c-bfbf33df755d.mp4

2. inspector panel drives size and alignment of rendered shape in thumbnail

https://user-images.githubusercontent.com/43151199/118962839-1a102300-b966-11eb-9cc3-54828b9c46ab.mp4

**Review by CertifiedRice — APPROVED** [review]

**Review by erikabar — CHANGES_REQUESTED** [review]
I am getting shader error when opening a project in URP: 

> Shader error in 'Hidden/Core/LensFlareDataDrivenPreview': Couldn't open include file 'Packages/com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl'. at line 26
> 
> Compiling Fragment program
> Platform defines: UNITY_ENABLE_REFLECTION_BUFFERS UNITY_USE_DITHER_MASK_FOR_ALPHABLENDED_SHADOWS UNITY_PBS_USE_BRDF1 UNITY_SPECCUBE_BOX_PROJECTION UNITY_SPECCUBE_BLENDING UNITY_ENABLE_DETAIL_NORMALMAP SHADER_API_DESKTOP UNITY_LIGHT_PROBE_PROXY_VOLUME UNITY_LIGHTMAP_RGBM_ENCODING
> Disabled keywords: FLARE_CIRCLE FLARE_POLYGON FLARE_INVERSE_SDF UNITY_NO_DXT5nm UNITY_FRAMEBUFFER_FETCH_AVAILABLE UNITY_ENABLE_NATIVE_SHADOW_LOOKUPS UNITY_METAL_SHADOWS_USE_POINT_FILTERING UNITY_NO_SCREENSPACE_SHADOWS UNITY_PBS_USE_BRDF2 UNITY_PBS_USE_BRDF3 UNITY_NO_FULL_STANDARD_SHADER UNITY_HARDWARE_TIER1 UNITY_HARDWARE_TIER2 UNITY_HARDWARE_TIER3 UNITY_COLORSPACE_GAMMA UNITY_HALF_PRECISION_FRAGMENT_SHADER_REGISTERS UNITY_LIGHTMAP_DLDR_ENCODING UNITY_LIGHTMAP_FULL_HDR UNITY_VIRTUAL_TEXTURING UNITY_PRETRANSFORM_TO_DISPLAY_ORIENTATION UNITY_ASTC_NORMALMAP_ENCODING SHADER_API_GLES30 UNITY_UNIFIED_SHADER_PRECISION_MODEL
>

**Comment by skhiat** [comment]
> I am getting shader error when opening a project in URP:
> 
> > Shader error in 'Hidden/Core/LensFlareDataDrivenPreview': Couldn't open include file 'Packages/com.unity.render-pipelines.high-definition/Runtime/ShaderLibrary/ShaderVariables.hlsl'. at line 26
> > Compiling Fragment program
> > Platform defines: UNITY_ENABLE_REFLECTION_BUFFERS UNITY_USE_DITHER_MASK_FOR_ALPHABLENDED_SHADOWS UNITY_PBS_USE_BRDF1 UNITY_SPECCUBE_BOX_PROJECTION UNITY_SPECCUBE_BLENDING UNITY_ENABLE_DETAIL_NORMALMAP SHADER_API_DESKTOP UNITY_LIGHT_PROBE_PROXY_VOLUME UNITY_LIGHTMAP_RGBM_ENCODING
> > Disabled keywords: FLARE_CIRCLE FLARE_POLYGON FLARE_INVERSE_SDF UNITY_NO_DXT5nm UNITY_FRAMEBUFFER_FETCH_AVAILABLE UNITY_ENABLE_NATIVE_SHADOW_LOOKUPS UNITY_METAL_SHADOWS_USE_POINT_FILTERING UNITY_NO_SCREENSPACE_SHADOWS UNITY_PBS_USE_BRDF2 UNITY_PBS_USE_BRDF3 UNITY_NO_FULL_STANDARD_SHADER UNITY_HARDWARE_TIER1 UNITY_HARDWARE_TIER2 UNITY_HARDWARE_TIER3 UNITY_COLORSPACE_GAMMA UNITY_HALF_PRECISION_FRAGMENT_SHADER_REGISTERS UNITY_LIGHTMAP_DLDR_ENCODING UNITY_LIGHTMAP_FULL_HDR UNITY_VIRTUAL_TEXTURING UNITY_PRETRANSFORM_TO_DISPLAY_ORIENTATION UNITY_ASTC_NORMALMAP_ENCODING SHADER_API_GLES30 UNITY_UNIFIED_SHADER_PRECISION_MODEL
> 
> also this error when creating a Lens Flare asset in URP:
> 
> > UnityException: FindShaderByName is not allowed to be called from a ScriptableObject constructor (or instance field initializer), call it in OnEnable instead. Called from ScriptableObject 'LensFlareDataSRPEditor'.
> > See "Script Serialization" page in the Unity Manual for further details.
> > UnityEngine.ResourcesAPI.FindShaderByName (System.String name) (at /Users/bokken/buildslave/unity/build/Runtime/Export/Resources/Resources.bindings.cs:65)
> > UnityEngine.Shader.Find (System.String name) (at /Users/bokken/buildslave/unity/build/Runtime/Export/Shaders/Shader.bindings.cs:31)
> > UnityEditor.Rendering.LensFlareDataSRPEditor..cctor () (at /Users/erikab/Documents/Github/Graphics/com.unity.render-pipelines.core/Editor/PostProcessing/LensFlareDataSRPEditor.cs:124)
> > Rethrow as TypeInitializationException: The type initializer for 'UnityEditor.Rendering.LensFlareDataSRPEditor' threw an exception.
> > UnityEditor.UIElements.InspectorElement+<>c__DisplayClass59_0.b__0 () (at /Users/bokken/buildslave/unity/build/ModuleOverrides/com.unity.ui/Editor/Inspector/InspectorElement.cs:625)
> > UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr, Boolean&) (at /Users/bokken/buildslave/unity/build/Modules/IMGUI/GUIUtility.cs:189)

[Fixed]

File: com.unity.render-pipelines.core/Editor/PostProcessing/LensFlareResource/TintedTexture.shader [hunk 1]
```diff
@@ -0,0 +1,61 @@
+Shader "Hidden/Core/TintedTexture"
```
Comment (RSlysz) [inline com.unity.render-pipelines.core/Editor/PostProcessing/LensFlareResource/TintedTexture.shader:1]: This file and the meta is not needed anymore. Please remove it :)

**Review by RSlysz — CHANGES_REQUESTED** [review]

**Review by iM0ve — APPROVED** [review]
Updated my review above. All issues fixed or addressed

**Review by RSlysz — APPROVED** [review]

**Review by eh-unity — APPROVED** [review]
Seems ok. Minor comments mostly about code readability for someone who doesn't know the implementation.

**Comment by RemyUnity** [comment]
Requesting a re-review from @erikabar or URP QA following Soufiane's fix : https://github.com/Unity-Technologies/Graphics/pull/4560#issuecomment-850321365

File: com.unity.render-pipelines.core/Editor/PostProcessing/LensFlareDataSRPEditor.cs [hunk 2]
```diff
@@ -299,6 +364,213 @@ int GetMultipleElementsCathegoryLines(SerializedProperty enabled, SerializedProp
...
+            float an = 2.0f * Mathf.PI * invSideCount;
+            float he = r * Mathf.Tan(0.5f * an);
+
+            float usedGradientPosition = Mathf.Clamp01((1.0f - edgeOffsetProp.floatValue) - 1e-6f);
+            if (type == SRPLensFlareType.Polygon)
+                usedGradientPosition = Mathf.Pow(usedGradientPosition + 1.0f, 5);
+
+            Vector4 flareData0 = LensFlareCommonSRP.GetFlareData0(Vector2.zero, Vector2.zero, Vector2.one, rotationProp.floatValue, 0f, 0f, Vector2.zero, false);
+
+            float cos0 = flareData0.x;
```
[the remark this answers is not in the document]
Comment (skhiat) [reply com.unity.render-pipelines.core/Editor/PostProcessing/LensFlareDataSRPEditor.cs:424]: More cos_theta, sin_theta.

**Review by erikabar — APPROVED** [review]
Tested URP, replicated all tests HDRP QA covered. In addition, tested removing array elements + undoing which resulted in ArgumentOutOfRange exception and the list is broken until you deselect the Lens Flare asset and select it again. This will be addressed in another PR.

**Pull request merged**